src/libbson/src/bson/bson-utf8.c

```diff
@@ -264,77 +376,104 @@ bson_utf8_escape_for_json (const char *utf8, /* IN */
    bson_unichar_t c;
    bson_string_t *str;
    bool length_provided = true;
+   ssize_t pos;
    const char *end;
 
    BSON_ASSERT (utf8);
 
-   str = bson_string_new (NULL);
-
    if (utf8_len < 0) {
       length_provided = false;
       utf8_len = strlen (utf8);
    }
 
+   if (utf8_len == 0) {
+      return bson_strdup ("");
+   }
+
    end = utf8 + utf8_len;
 
-   while (utf8 < end) {
+   str = bson_string_alloc (utf8_len);
+
+   pos = 0;
+
+   do {
       uint8_t mask;
       uint8_t length_of_char;
 
+      c = (unsigned char) utf8[pos];
+      if (!_is_special_char (c)) {
+         // Normal character, no need to do anything besides iterate
+         // Copy rest of the string if we reach the end
+         pos++;
+         utf8_len--;
+         if (utf8_len == 0) {
+            bson_string_append_ex (str, utf8, pos);
+            break;
+         }
+
+         continue;
+      }
+
+      // Reached a special character. Copy over all of normal characters
+      // we have passed so far
+      if (pos) {
```

Comment:
```suggestion
      if (pos > 0) {
```
Avoid treating integers as booleans.